zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/internal/JsonReaders.java

```diff
@@ -25,85 +25,77 @@
-   * }</pre>
-   *
-   * <p>You can advance to the field directly.
+   * Navigates to a field of a JSON-serialized object. For example,
    *
    * <pre>{@code
-   * JsonReader status = enterPath(JsonReader.of(body.source()), "message", "status");
+   * JsonReader status = enterPath(JsonAdapters.jsonParser(stream), "message", "status");
    * if (status != null) throw new IllegalStateException(status.nextString());
    * }</pre>
    */
   @Nullable
-  public static JsonReader enterPath(JsonReader reader, String path1, String path2)
+  public static JsonParser enterPath(JsonParser parser, String path1, String path2)
       throws IOException {
-    return enterPath(reader, path1) != null ? enterPath(reader, path2) : null;
+    return enterPath(parser, path1) != null ? enterPath(parser, path2) : null;
   }
 
   @Nullable
-  public static JsonReader enterPath(JsonReader reader, String path) throws IOException {
-    try {
-      if (reader.peek() != JsonReader.Token.BEGIN_OBJECT) return null;
-    } catch (EOFException e) {
-      return null;
-    }
-    reader.beginObject();
-    while (reader.hasNext()) {
-      if (reader.nextName().equals(path) && reader.peek() != JsonReader.Token.NULL) {
-        return reader;
+  public static JsonParser enterPath(JsonParser parser, String path) throws IOException {
+    if (!parser.isExpectedStartObjectToken()) return null;
+    JsonToken value;
+    while ((value = parser.nextValue()) != JsonToken.END_OBJECT) {
+      if (value == null) {
+        // End of input.
+        throw new IOException("End of input while parsing object.");
+      }
+      if (parser.getCurrentName().equals(path) && value != JsonToken.VALUE_NULL) {
+        return parser;
       } else {
-        reader.skipValue();
+        parser.skipChildren();
       }
     }
-    reader.endObject();
     return null;
   }
 
-  public static List<String> collectValuesNamed(JsonReader reader, String name) throws IOException {
+  public static List<String> collectValuesNamed(JsonParser parser, String name) throws IOException {
     Set<String> result = new LinkedHashSet<>();
-    visitObject(reader, name, result);
+    visitObject(parser, name, result);
     return new ArrayList<>(result);
   }
 
-  static void visitObject(JsonReader reader, String name, Set<String> result) throws IOException {
-    reader.beginObject();
-    while (reader.hasNext()) {
-      if (reader.nextName().equals(name)) {
-        result.add(reader.nextString());
+  static void visitObject(JsonParser parser, String name, Set<String> result) throws IOException {
+    if (!parser.isExpectedStartObjectToken()) {
+      throw new IOException("Expecting object start, got " + parser.currentToken());
+    }
+    JsonToken value;
+    while ((value = parser.nextValue()) != JsonToken.END_OBJECT) {
+      if (value == null) {
+        throw new IOException("End of input while parsing object.");
```

Comment:
I think the IOExceptions here could be turned to null returns possibly log.fine guarded if important. Starting to throw when wasn't before could break things.

Reply:
See above - moshi should have been throwing JsonDataException in the old code - I'm surprised that exception is not an IOException though since moshi's methods declare `throws IOException` too

Reply:
anyway I wrote this code :) it wasn't planned to have multiple layers of catch. it would be a heck of a lot nicer to not have to eyeball if correct or not and instead just return null.

Reply:
ex if you look it does things like this

```
    return enterPath(reader, path1) != null ? enterPath(reader, path2) : null;
```

if you feel strongly about making it throw that's ok just it is tiring to fix little exceptions here and there.

Reply:
Ah ok it does make more sense here. It means we're returning `null` where we used to throw, seems safe here though.